File: packages/dotnet/TezosPayments/Serialization/PaymentSerializer.cs
using System.Text.Json;
using TezosPayments.Models;

namespace TezosPayments.Serialization;

public class JsonPaymentSerializer : IPaymentSerializer<string>
{
    public JsonSerializerOptions JsonSerializerOptions = new()
    {
        DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull
    };

    public virtual string Serialize(Payment payment)
        => JsonSerializer.Serialize(MapPaymentToSerializedPayment(payment), JsonSerializerOptions);

    public virtual Task<string> SerializeAsync(Payment payment) => Task.FromResult(Serialize(payment));

    protected virtual SerializedPayment MapPaymentToSerializedPayment(Payment payment) => new()
    {
        Id = payment.Id,
        Amount = payment.Amount,
        Created = new DateTimeOffset(payment.Created).ToUnixTimeMilliseconds(),
        Data = payment.Data,
        Asset = payment.Asset != null ? MapPaymentAssetToSerializedPaymentAsset(payment.Asset) : null,
        SuccessUrl = payment.SuccessUrl?.AbsoluteUri,
        CancelUrl = payment.CancelUrl?.AbsoluteUri,
        Expired = payment.Expired != null ? new DateTimeOffset(payment.Expired.Value).ToUnixTimeMilliseconds() : null,
        Signature = payment.Signature != null ? MapPaymentSignatureToSerializedPaymentSignature(payment.Signature) : null
    };

    protected virtual SerializedPaymentAsset MapPaymentAssetToSerializedPaymentAsset(PaymentAsset paymentAsset)
        => new(paymentAsset.Address, paymentAsset.Decimals, paymentAsset.Id);

    protected virtual SerializedPaymentSignature MapPaymentSignatureToSerializedPaymentSignature(PaymentSignature paymentSignature)
        => new(paymentSignature.Contract, paymentSignature.Client);
}

File: packages/js/common/src/serialization/payment/paymentSerializer.ts
import type {
  Payment, PaymentAsset, PaymentSignature,
  SerializedPayment, SerializedPaymentAsset, SerializedPaymentSignature
} from '../../models';
import { Base64Serializer } from '../base64';
import { serializedPaymentFieldTypes } from './serializedPaymentFieldTypes';

export class PaymentSerializer {
  protected static readonly serializedPaymentBase64Serializer = new Base64Serializer<SerializedPayment>(
    serializedPaymentFieldTypes
  );

  serialize(payment: Payment): string | null {
    try {
      const serializedPayment = this.mapPaymentToSerializedPayment(payment);
      return PaymentSerializer.serializedPaymentBase64Serializer.serialize(serializedPayment);
    }
    catch {
      return null;
    }
  }

  protected mapPaymentToSerializedPayment(payment: Payment): SerializedPayment {
    return {
      i: payment.id,
      a: payment.amount.toString(),
      t: payment.targetAddress,
      as: payment.asset ? this.mapPaymentAssetToSerializedPaymentAsset(payment.asset) : undefined,
      d: payment.data,
      su: payment.successUrl?.toString(),
      cu: payment.cancelUrl?.toString(),
      c: payment.created.getTime(),
      e: payment.expired?.getTime(),
      s: this.mapPaymentSignatureToSerializedPaymentSignature(payment.signature)
    };
  }

  protected mapPaymentAssetToSerializedPaymentAsset(paymentAsset: PaymentAsset): SerializedPaymentAsset {
    return {
      a: paymentAsset.address,
      d: paymentAsset.decimals,
      i: paymentAsset.id !== null ? paymentAsset.id : undefined
    };
  }

  protected mapPaymentSignatureToSerializedPaymentSignature(paymentSignature: PaymentSignature): SerializedPaymentSignature {
    return {
      k: paymentSignature.signingPublicKey,
      c: paymentSignature.contract,
      cl: paymentSignature.client
    };
  }
}
